ops/main.py

```diff
@@ -169,6 +169,22 @@ def main(charm_class):
     # TODO: For Windows, when symlinks are used, this is not a valid
     #       method of getting an event name (see LP: #1854505).
     juju_exec_path = Path(sys.argv[0])
+    has_dispatch = juju_exec_path.name == 'dispatch'
+    if has_dispatch:
+        # The executable was 'dispatch', which means the actual hook we want to
+        # run needs to be looked up in the JUJU_DISPATCH_PATH env var, where it
+        # should be a path relative to the charm directory (the directory that
+        # holds `dispatch`). If that path actually exists, we need to exec to
+        # it instead of continuing.
+        #
+        # We don't look at the existence of JUJU_DISPATCH_PATH because we don't
+        # unset it on exec'ing into the hook, and it could be a symlink back to
+        # ourselves.
+        dispatch_path = Path(os.environ['JUJU_DISPATCH_PATH'])
+        juju_exec_path = juju_exec_path.parent / dispatch_path
+        if juju_exec_path.exists():
+            sys.argv[0] = str(juju_exec_path)
+            os.execv(sys.argv[0], sys.argv)
     juju_event_name = juju_exec_path.name.replace('-', '_')
```

Comment:
You need to interpret JUJU_HOOK_NAME or JUJU_DISPATCH_PATH to get the 'juju_event_name' if the juju_exec_path is 'dispatch'. Because otherwise we'll be trying to fire the Charm.on.dispatch I think from how I read this.

Reply:
still need to do this